tests/rust/src/bin/poll_oneoff_stdio.rs

```diff
@@ -0,0 +1,137 @@
+use std::collections::HashMap;
+use std::mem::MaybeUninit;
+use wasi_tests::{assert_errno, STDERR_FD, STDIN_FD, STDOUT_FD};
+
+const TIMEOUT: u64 = 200_000_000u64; // 200 milliseconds, required to satisfy slow execution in CI
+const CLOCK_ID: wasi::Userdata = 0x0123_45678;
+const STDIN_ID: wasi::Userdata = 0x8765_43210;
+
+unsafe fn poll_oneoff_impl(r#in: &[wasi::Subscription]) -> Result<Vec<wasi::Event>, wasi::Errno> {
+    let mut out: Vec<wasi::Event> = Vec::new();
+    out.resize_with(r#in.len(), || {
+        MaybeUninit::<wasi::Event>::zeroed().assume_init()
+    });
+    let size = wasi::poll_oneoff(r#in.as_ptr(), out.as_mut_ptr(), r#in.len())?;
+    out.truncate(size);
+    Ok(out)
+}
+
+unsafe fn test_stdin_read() {
+    let clock = wasi::SubscriptionClock {
+        id: wasi::CLOCKID_MONOTONIC,
+        timeout: TIMEOUT,
+        precision: 0,
+        flags: 0,
+    };
+    let fd_readwrite = wasi::SubscriptionFdReadwrite {
+        file_descriptor: STDIN_FD,
+    };
+    // Either stdin can be ready for reading, or this poll can timeout.
+    let r#in = [
+        wasi::Subscription {
+            userdata: CLOCK_ID,
+            u: wasi::SubscriptionU {
+                tag: wasi::EVENTTYPE_CLOCK.raw(),
+                u: wasi::SubscriptionUU { clock },
+            },
+        },
+        wasi::Subscription {
+            userdata: STDIN_ID,
+            u: wasi::SubscriptionU {
+                tag: wasi::EVENTTYPE_FD_READ.raw(),
+                u: wasi::SubscriptionUU {
+                    fd_read: fd_readwrite,
+                },
+            },
+        },
+    ];
+    let out = poll_oneoff_impl(&r#in).unwrap();
+    // The result should be either a timeout, or that stdin is ready for reading.
+    // Both are valid behaviors that depend on the test environment.
+    assert!(out.len() >= 1, "stdin read should return at least 1 event");
+    for event in out {
+        if event.type_ == wasi::EVENTTYPE_CLOCK {
+            assert_errno!(event.error, wasi::ERRNO_SUCCESS);
+            assert_eq!(
+                event.userdata, CLOCK_ID,
+                "the event.userdata should contain CLOCK_ID",
+            );
+        } else if event.type_ == wasi::EVENTTYPE_FD_READ {
+            assert_errno!(event.error, wasi::ERRNO_SUCCESS);
```

Comment:
any idea about this? https://github.com/yamt/toywasm/blob/7e46dee9758f0f72a7d5f390790f35363b8feb5d/test/wasmtime-wasi-tests-blacklist.txt#L24-L30

Reply:
FYI, i ended up with this: https://github.com/yamt/toywasm/commit/570e6706312cbdccb9d5268347186b441a24444b